default/AI/MilitaryAI.py

```diff
@@ -16,6 +17,34 @@
 totMilRating = 0
 num_milships = 0
 verbose_mil_reporting = False
+best_ship_rating_cache = {}  # indexed by turn, value is rating of that turn
+
+
+def cur_best_mil_ship_rating(include_designs=False):
+    """Find the best military ship we have available in this turn and return its rating.
+
+    :param include_designs: toggles if available designs are considered or only existing ships
+    :return: float: rating of the best ship
+    """
+    current_turn = fo.currentTurn()
+    if current_turn in best_ship_rating_cache:
+        return best_ship_rating_cache[current_turn]
+    best_rating = 0.001
+    universe = fo.getUniverse()
+    for fleet_id in FleetUtilsAI.get_empire_fleet_ids_by_role(AIFleetMissionType.FLEET_MISSION_MILITARY):
```

Comment:
May be move code to find ships outside, do you common with generators?

Make something like that in FleetUtilsAI
```python
def iterate_ships(fleet_list):
    universe = fo.getUniverse()
    for fleet_id in fleet_list:
        fleet = universe.getFleet(fleet_id)
        for ship_id in fleet.shipIDs:
            ship = universe.getShip(ship_id)
            if ship:
                yield ship
```

And use it like this:

```python
def best_rating(current_rating, ship):
    ship_info = [(ship.id, ship.designID, ship.speciesName)]
    return max(current_rating, foAI.foAIstate.rate_psuedo_fleet(ship_info=ship_info)['overall'])

best_rating = reduce(best_rating, iterate_ships(fleet_list), 0.001)  # max does not work with empty list